File: dsa/easy/Tournament Winner/tournament_winner.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

std::string tournamentWinner(std::vector<std::vector<std::string>> competitions,
                             std::vector<int> results) {
  std::unordered_map<std::string, int> scores;
  for (int i = 0; i < competitions.size(); i++) {
    std::string curr_winner = competitions[i][results[i] == 1 ? 0 : 1];
    if (scores.find(curr_winner) == scores.end()) {
      scores[curr_winner] = 0;
    }
    scores[curr_winner] += 3;
  }
  std::string winner = "";
  int max = 0;
  for (auto it = scores.begin(); it != scores.end(); it++) {
    // only if it's greater than the current max
    if (it->second > max) {
      max = it->second;
      winner = it->first;
    }
  }
  return winner;
}

File: dsa/easy/Tournament Winner/tournament_winner.go
package main

const HOME_TEAM_WON = 1

func TournamentWinner(competitions [][]string, results []int) string {
	winner_board := make(map[string]int)
	for i := 0; i < len(competitions); i++ {
		// by default we set as the away team the winner
		targetIdx := 1
		if results[i] == HOME_TEAM_WON {
			targetIdx = 0
		}
		curr_winner := competitions[i][targetIdx]
		winner_board[curr_winner] += 3
	}
	winner := ""
	curr_max := 0
	for k, v := range winner_board {
		if v > curr_max {
			curr_max = v
			winner = k
		}
	}
	return winner
}
